staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -17,77 +17,199 @@ limitations under the License.
 package log
 
 import (
+	"errors"
 	"fmt"
 	"io"
 	"strings"
+	"sync"
+	"time"
 
+	"github.com/golang/glog"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
 	auditinternal "k8s.io/apiserver/pkg/apis/audit"
 	"k8s.io/apiserver/pkg/audit"
 )
 
 const (
+	// ModeBlocking indicates that the backend should process events directly,
+	// Serialize event, then write it to logs. This mode spend more time.
+	ModeBlocking = "blocking"
+	// ModeBuffered indicates that the audit backend should buffer audit events
+	// internally. If the buffer is full, audit event will be dropped, so no
+	// block event producer.
+	ModeBuffered = "buffered"
+)
+
+// AllowedModes is the modes known by this webhook.
+var AllowedModes = []string{
+	ModeBlocking,
+	ModeBuffered,
+}
+
+const (
 	// FormatLegacy saves event in 1-line text format.
 	FormatLegacy = "legacy"
 	// FormatJson saves event in structured json format.
 	FormatJson = "json"
+
+	// TODO: make this configurable
+	// Number of workers that concurrently write events.
+	workers = 5
+	// Buffer up to 1000 events.
+	defaultBufferSize = 1000
 )
 
+// The plugin name reported in error metrics.
+const pluginName = "log"
+
 // AllowedFormats are the formats known by log backend.
 var AllowedFormats = []string{
 	FormatLegacy,
 	FormatJson,
 }
 
-type backend struct {
+func NewBackend(out io.Writer, format, mode string, groupVersion schema.GroupVersion) (audit.Backend, error) {
+	switch mode {
+	case ModeBuffered:
+		return newBufferedBackend(out, format, groupVersion), nil
+	case ModeBlocking:
+		return newDirectBackend(out, format, groupVersion), nil
+	default:
+		return nil, fmt.Errorf("log mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+}
+
+type bufferedBackend struct {
 	out          io.Writer
 	format       string
 	groupVersion schema.GroupVersion
+	buffer       chan *auditinternal.Event
+	wg           sync.WaitGroup
 }
 
-var _ audit.Backend = &backend{}
+var _ audit.Backend = &bufferedBackend{}
 
-func NewBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
-	return &backend{
+func newBufferedBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
+	return &bufferedBackend{
 		out:          out,
 		format:       format,
 		groupVersion: groupVersion,
+		buffer:       make(chan *auditinternal.Event, defaultBufferSize),
+		wg:           sync.WaitGroup{},
 	}
 }
 
-func (b *backend) ProcessEvents(events ...*auditinternal.Event) {
+func (b *bufferedBackend) ProcessEvents(events ...*auditinternal.Event) {
```

Comment:
audit events are reused. If you want to hold on to them, you need to deep copy it.

https://github.com/kubernetes/kubernetes/blob/f369c1af596ecd8b5b54916a0cbbface8f422121/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L333-L335

Reply:
I did see this deepCopy, but didnot find where it is reused.

Reply:
@ericchiang Thanks, after re-read the caller code, I see.